### PR TITLE
[SPARK-7180][SPARK-8090][SPARK-8091] Fix a number of SerializationDebugger bugs and limitations

### PR DESCRIPTION
**Pull request by tdas**
This PR solves three SerializationDebugger issues. 
- SPARK-7180 - SerializationDebugger fails with ArrayOutOfBoundsException
- SPARK-8090 - SerializationDebugger does not handle classes with writeReplace correctly
- SPARK-8091 - SerializationDebugger does not handle classes with writeObject method

The solutions for each are explained as follows
- SPARK-7180 - The wrong slot desc was used for getting the value of the fields in the object being tested. 
- SPARK-8090 - Test the type of the replaced object.
- SPARK-8091 - Use a dummy ObjectOutputStream to collect all the objects written by the writeObject() method, and then test those objects as usual.

I also added more tests in the testsuite to increase code coverage. For example, added tests for cases where there are not serializability issues.

### CODE REVIEW

**Comment by tdas** [comment]
@andrewor14 Please take a look. 
I am still adding more comments, so this is sort-of WIP.

**Comment by tdas** [comment]
@rxin Take a look. Adding comments, so slightly WIP.

File: core/src/main/scala/org/apache/spark/serializer/SerializationDebugger.scala [hunk 1]
```diff
@@ -145,17 +145,25 @@ private[spark] object SerializationDebugger extends Logging {
...
+          val childStack = visitSerializableWithWriteObjectMethod(finalObj, slotDesc, stack)
+          if (childStack.nonEmpty) {
+            return childStack
+          }
         } else {
           val fields: Array[ObjectStreamField] = slotDesc.getFields
           val objFieldValues: Array[Object] = new Array[Object](slotDesc.getNumObjFields)
           val numPrims = fields.length - objFieldValues.length
-          desc.getObjFieldValues(finalObj, objFieldValues)
+          slotDesc.getObjFieldValues(finalObj, objFieldValues)
```
Comment (tdas) [inline core/src/main/scala/org/apache/spark/serializer/SerializationDebugger.scala]: This is the bug fix for SPARK-7180

File: core/src/main/scala/org/apache/spark/serializer/SerializationDebugger.scala [hunk 2]
```diff
@@ -145,17 +145,25 @@ private[spark] object SerializationDebugger extends Logging {
       // An object contains multiple slots in serialization.
       // Get the slots and visit fields in all of them.
       val (finalObj, desc) = findObjectAndDescriptor(o)
+
+      if (!finalObj.eq(o)) {
+        return visit(finalObj, s"writeReplace data (class: ${finalObj.getClass.getName})" :: stack)
+      }
```
Comment (tdas) [inline core/src/main/scala/org/apache/spark/serializer/SerializationDebugger.scala]: This is the fix for SPARK-8090

File: core/src/main/scala/org/apache/spark/serializer/SerializationDebugger.scala [hunk 3]
```diff
@@ -169,12 +177,39 @@ private[spark] object SerializationDebugger extends Logging {
             }
             j += 1
           }
-
         }
         i += 1
       }
       return List.empty
     }
+
```
Comment (tdas) [inline core/src/main/scala/org/apache/spark/serializer/SerializationDebugger.scala]: This is the fix for SPARK-8191

File: core/src/main/scala/org/apache/spark/serializer/SerializationDebugger.scala [hunk 4]
```diff
@@ -145,17 +145,25 @@ private[spark] object SerializationDebugger extends Logging {
...
       val slotDescs = desc.getSlotDescs
       var i = 0
       while (i < slotDescs.length) {
         val slotDesc = slotDescs(i)
         if (slotDesc.hasWriteObjectMethod) {
-          // TODO: Handle classes that specify writeObject method.
+          val childStack = visitSerializableWithWriteObjectMethod(finalObj, slotDesc, stack)
+          if (childStack.nonEmpty) {
+            return childStack
+          }
```
Comment (tdas) [inline core/src/main/scala/org/apache/spark/serializer/SerializationDebugger.scala]: This is the fix for SPARK-8191

File: core/src/main/scala/org/apache/spark/serializer/SerializationDebugger.scala [hunk 5]
```diff
@@ -169,12 +177,39 @@ private[spark] object SerializationDebugger extends Logging {
...
-
         }
         i += 1
       }
       return List.empty
     }
+
+    private def visitSerializableWithWriteObjectMethod(
+        o: Object, slotDesc: ObjectStreamClass, stack: List[String]): List[String] = {
+      println(">>> processing serializable with writeObject" + o)
```
Comment (andrewor14) [inline core/src/main/scala/org/apache/spark/serializer/SerializationDebugger.scala]: stray println

File: core/src/main/scala/org/apache/spark/serializer/SerializationDebugger.scala [hunk 6]
```diff
@@ -169,12 +177,39 @@ private[spark] object SerializationDebugger extends Logging {
...
           }
-
         }
         i += 1
       }
       return List.empty
     }
+
+    private def visitSerializableWithWriteObjectMethod(
+        o: Object, slotDesc: ObjectStreamClass, stack: List[String]): List[String] = {
```
Comment (andrewor14) [inline core/src/main/scala/org/apache/spark/serializer/SerializationDebugger.scala]: can you use this style:
```
private def theMethod(
    o: Object,
    slotDesc: ObjectStreamClass,
    stack: List[String]): List[String] = {
  ...
}
```
It would also be good to have a java doc that explains what the parameters are, especially since some of them have very terse names e.g. `o`

File: core/src/main/scala/org/apache/spark/serializer/SerializationDebugger.scala (hunk 5 above)
Comment (tdas) [inline core/src/main/scala/org/apache/spark/serializer/SerializationDebugger.scala]: Already removed, waiting for more feedback to accumulate before pushing.

File: core/src/main/scala/org/apache/spark/serializer/SerializationDebugger.scala [hunk 7]
```diff
@@ -220,6 +255,27 @@ private[spark] object SerializationDebugger extends Logging {
...
   }
 
+  /** An output stream that emulates /dev/null */
+  private class NullOutputStream extends OutputStream {
+    override def write(b: Int) { }
+  }
+
+  /**
+   * A dummy [[ObjectOutputStream]] that saves the list of objects written to it and returns
+   * them through `outputArray`.
```
Comment (andrewor14) [inline core/src/main/scala/org/apache/spark/serializer/SerializationDebugger.scala]: can you expand on this a little? e.g. why pass it null output stream, what is it for, why we need to extend `ObjectOutputStream` etc.

File: core/src/main/scala/org/apache/spark/serializer/SerializationDebugger.scala (hunk 6 above)
Comment (tdas) [inline core/src/main/scala/org/apache/spark/serializer/SerializationDebugger.scala]: Yep, adding that.

File: core/src/test/scala/org/apache/spark/serializer/SerializationDebuggerSuite.scala [hunk 8]
```diff
@@ -106,6 +106,48 @@ class SerializationDebuggerSuite extends SparkFunSuite with BeforeAndAfterEach {
...
+      new SerializableClassWithWriteReplace(
+        new ExternalizableClass(
+          new SerializableSubclass(
+            new SerializableArray(
+              Array(new SerializableClass1, new SerializableClass2(new NotSerializable))
+            ))))
+    )
+    assert(s.nonEmpty)
+    assert(s.head.contains("NotSerializable"))
+  }
```
Comment (andrewor14) [inline core/src/test/scala/org/apache/spark/serializer/SerializationDebuggerSuite.scala]: yay!

File: core/src/main/scala/org/apache/spark/serializer/SerializationDebugger.scala (hunk 2 above)
Comment (andrewor14) [inline core/src/main/scala/org/apache/spark/serializer/SerializationDebugger.scala]: need comment

**Comment by andrewor14** [comment]
This looks great. I am looking forward to your land mines of comments...

**Comment by tdas** [comment]
@andrewor14 How about them comments? Also added more tests. 93% line coverage.

screen shot 2015-06-03 at 6 21 33 pm

File: core/src/main/scala/org/apache/spark/serializer/SerializationDebugger.scala [hunk 9]
```diff
@@ -62,7 +62,7 @@ private[spark] object SerializationDebugger extends Logging {
    *
    * It does not yet handle writeObject override, but that shouldn't be too hard to do either.
    */
-  def find(obj: Any): List[String] = {
+  private[serializer] def find(obj: Any): List[String] = {
```
Comment (tdas) [inline core/src/main/scala/org/apache/spark/serializer/SerializationDebugger.scala]: so that everyone uses `improveException` which is safe to call.

File: core/src/main/scala/org/apache/spark/serializer/SerializationDebugger.scala [hunk 10]
```diff
@@ -125,6 +125,13 @@ private[spark] object SerializationDebugger extends Logging {
       return List.empty
     }
 
+    /**
+     * Visit an externalizable object.
+     * Since writeExternal() can choose add arbitrary objects at the time of serialization,
+     * the only way to capture all the objects it will serialize is by using a
+     * dummy ObjectOutput object that captures all the inner objects, and then visit all the
```
Comment (andrewor14) [inline core/src/main/scala/org/apache/spark/serializer/SerializationDebugger.scala]: This is a really long sentence and fairly low level. How about:
```
Since writeExternal() can choose add arbitrary objects at the time of serialization,
the only way to capture all the objects it will serialize is to manually write the object
into our own dummy ObjectOutput that collects the relevant fields.
```

Comment (tdas) [inline core/src/main/scala/org/apache/spark/serializer/SerializationDebugger.scala]: Okay, I can shorten it for this, as well as for the new visit function.

File: core/src/main/scala/org/apache/spark/serializer/SerializationDebugger.scala [hunk 11]
```diff
@@ -145,17 +151,50 @@ private[spark] object SerializationDebugger extends Logging {
...
       val (finalObj, desc) = findObjectAndDescriptor(o)
+
+      // If the object has been replaced using writeReplace(),
+      // then call visit() on it again to test its type again.
+      if (!finalObj.eq(o)) {
+        return visit(finalObj, s"writeReplace data (class: ${finalObj.getClass.getName})" :: stack)
+      }
+
+      // Every class is associated with one or more "slots", each slot is related to the parent
+      // classes of this class. These slots are used by the ObjectOutputStream
```
Comment (andrewor14) [inline core/src/main/scala/org/apache/spark/serializer/SerializationDebugger.scala]: grammar nit: each slot refers to the parent class of this class

File: core/src/main/scala/org/apache/spark/serializer/SerializationDebugger.scala [hunk 12]
```diff
@@ -145,17 +151,50 @@ private[spark] object SerializationDebugger extends Logging {
...
+      // serialization code to recursively serialize the fields of an object and
+      // its parent classes. For example, if there are the following classes.
+      //
+      //     class ParentClass(parentField: Int)
+      //     class ChildClass(childField: Int) extends ParentClass(1)
+      //
+      // Then serializing the an object Obj of type ChildClass requires for serializing the fields
```
Comment (andrewor14) [inline core/src/main/scala/org/apache/spark/serializer/SerializationDebugger.scala]: requires for -> requires first

File: core/src/main/scala/org/apache/spark/serializer/SerializationDebugger.scala [hunk 13]
```diff
@@ -145,17 +151,50 @@ private[spark] object SerializationDebugger extends Logging {
...
+      // of ParentClass (that is, parentField), and then serializing the fields of ChildClass
+      // (that is, childField). Correspondingly, there will be two slots related to this object:
+      //
+      // 1. ParentClass slot, which will be used to serialize parentField of Obj
+      // 2. ChildClass slot, which will be used to serialize childField fields of Obj
+      //
+      // The following code uses the description of each slot related to the object to get the
+      // corresponding field objects and then visits them to test their serializability.
```
Comment (andrewor14) [inline core/src/main/scala/org/apache/spark/serializer/SerializationDebugger.scala]: this sentence is so windy. How about just
```
// This code uses the description of each slot to find the fields to visit
```

File: core/src/main/scala/org/apache/spark/serializer/SerializationDebugger.scala [hunk 14]
```diff
@@ -169,18 +208,53 @@ private[spark] object SerializationDebugger extends Logging {
...
-
         }
         i += 1
       }
       return List.empty
     }
+
+    /**
+     * Visit a serializable object which has the writeObject() defined.
+     * Since writeObject() can choose add arbitrary objects at the time of serialization,
```
Comment (andrewor14) [inline core/src/main/scala/org/apache/spark/serializer/SerializationDebugger.scala]: can choose to

File: core/src/main/scala/org/apache/spark/serializer/SerializationDebugger.scala [hunk 15]
```diff
@@ -125,6 +125,12 @@ private[spark] object SerializationDebugger extends Logging {
       return List.empty
     }
 
+    /**
+     * Visit an externalizable object.
+     * Since writeExternal() can choose add arbitrary objects at the time of serialization,
```
Comment (andrewor14) [inline core/src/main/scala/org/apache/spark/serializer/SerializationDebugger.scala]: can choose to

File: core/src/main/scala/org/apache/spark/serializer/SerializationDebugger.scala [hunk 16]
```diff
@@ -169,18 +208,53 @@ private[spark] object SerializationDebugger extends Logging {
...
         i += 1
       }
       return List.empty
     }
+
+    /**
+     * Visit a serializable object which has the writeObject() defined.
+     * Since writeObject() can choose add arbitrary objects at the time of serialization,
+     * the only way to capture all the objects it will serialize is by using a
+     * dummy ObjectOutputStream that collects all the relevant fields for further testing.
```
Comment (andrewor14) [inline core/src/main/scala/org/apache/spark/serializer/SerializationDebugger.scala]: maybe it's worth mentioning that this is the similar to how externalizable objects are visited?

File: core/src/main/scala/org/apache/spark/serializer/SerializationDebugger.scala [hunk 17]
```diff
@@ -169,18 +208,53 @@ private[spark] object SerializationDebugger extends Logging {
...
+      try {
+        innerObjectsCatcher.writeObject(o)
+      } catch {
+        case io: IOException =>
+          notSerializableFound = true
+      }
+
+      // If something was not serializable, then visit the captured objects.
+      // Otherwise, all the captured objects are safely serializable, so no need to visit them.
+      // As an optimization, just added them to the visited list.
```
Comment (andrewor14) [inline core/src/main/scala/org/apache/spark/serializer/SerializationDebugger.scala]: I think this could just be:
```
// Do not bother visiting the captured objects if everything is serializable
```

Comment (andrewor14) [inline core/src/main/scala/org/apache/spark/serializer/SerializationDebugger.scala]: then down there in L248 you can optionally add the optimization comment (though I don't think it's necessary)

File: streaming/src/main/scala/org/apache/spark/streaming/StreamingContext.scala [hunk 18]
```diff
@@ -549,8 +549,8 @@ class StreamingContext private[streaming] (
         case e: NotSerializableException =>
           throw new NotSerializableException(
             "DStream checkpointing has been enabled but the DStreams with their functions " +
-              "are not serializable\nSerialization stack:\n" +
-              SerializationDebugger.find(checkpoint).map("\t- " + _).mkString("\n")
+              "are not serializable\n" +
+              SerializationDebugger.improveException(checkpoint, e).getMessage()
```
Comment (andrewor14) [inline streaming/src/main/scala/org/apache/spark/streaming/StreamingContext.scala]: this is a great change!

**Comment by andrewor14** [comment]
I left a bunch of english nits, but this is mergeable as is. LGTM.

**Comment by JoshRosen** [comment]
Bump @tdas @andrewor14, do we want to merge this now or should we wait for the comment changes?  We may want to retest with Jenkins first, just in case there's any new style rules that this might fail.

**Comment by andrewor14** [comment]
retest this please

**Comment by JoshRosen** [comment]
Gentle bump again: do we want to merge this as-is and fix the comments later?

**Comment by andrewor14** [comment]
@tdas is working on this right now I believe

File: core/src/main/scala/org/apache/spark/serializer/SerializationDebugger.scala (hunk 17 above)
Comment (tdas) [inline core/src/main/scala/org/apache/spark/serializer/SerializationDebugger.scala]: i think its weird to comment on the negative case ("do not visit") when the immediately below code is the positive case.

Comment (andrewor14) [inline core/src/main/scala/org/apache/spark/serializer/SerializationDebugger.scala]: it doesn't really matter as long as we get the message across... if you prefer we can rephrase this as "Only visit the captured objects if everything is serializable"

**Comment by tdas** [comment]
@andrewor14 This is good to go.

**Comment by andrewor14** [comment]
Ok, merging into master.

**Comment by andrewor14** [comment]
I also merged this into 1.4.1.
